GT Modular Armor

FourIsTheNumber opened the pull request:
IC2 is very old and ARR, and GTNH is slowly pushing towards obsoleting and removing it. Part of this effort is replacing the IC2 armor types, such as IC2 jetpacks, nano armor, quantum armor, gravichestplate (reliant on IC2).

To this end, this pr introduces Gregtech's own modular armor system (and an armor behavior system). It is composed of three parts: the frame, the energy core, and augments. 
Energy cores control the energy tier, maximum energy, and which augment tiers may be equipped.
Frames add slots to the armor from one of the three categories (Protection, Movement, and Utility). They may also add Prismatic slots which can hold any augment.
Augments provide all the varying effects. 
Some frames and cores may also provide certain effects.

 In HV, you will be able to construct the full set of armor and the first energy core/frames. Throughout progression, you can interchange components/core/frames at will.

## Cores:

-Energy Core (HV)
-Nano Core (IV)
-Living Core (ZPM)
-Singularity Core (UHV)

## Frames (Protection, Movement, Utility, Prismatic):
### Tier 1
-Light Frame (1, 2, 0, 0)
-Medium Frame (1, 1, 1, 0)
-Heavy Frame (2, 0, 1, 0)

### Tier 2
-Nimble Frame (1, 3, 1, 0)
-Adaptive Frame (2, 1, 1, 1)
-Tough Frame (3, 0, 2, 0)

### Tier 3
-Lightning Frame (1, 4, 1, 1)
-Morphic Frame (2, 1, 2, 2)
-Bulwark Frame (4, 1, 1, 1)

## Augments:

### Energy Core
-Hazmat (All Pieces, Protection)
-Space Suit (All Pieces, Protection)
-Night Vision (Helmet, Utility)
-Goggles of Revealing (Helmet, Utility)
-Water Breathing (Helmet, Utility)
-Jetpack (Chestplate, Movement)
-Apiarist (Leggings, Utility)
-Swim Speed (Boots, Movement)

### Nano Core
-Jetpack Perfect Hover (Chestplate, Movement)
-Fire Immunity (Leggings, Protection)
-Knockback Resistance (Leggings, Protection)
-Fall Protection (Boots, Protection)
-Step Assist (Boots, Movement)
-Omni Movement (Boots, Movement)
-Speed Boost (Boots, Movement, Stackable)
-Jump Boost (Boots, Movement, Stackable)

### Living Core
-Creative Flight (Chestplate, Movement)
-Inertia Canceling (Chestplate, Movement)
-Force Field (Chestplate, Protection)
-Teleporter (Leggings, Utility)

Note: No augments currently require the singularity core, but it will add infinite energy to the armor.

FourIsTheNumber committed:
Add some armor classes

FourIsTheNumber committed:
Add a prototype of the modification table and start messing with nbt

FourIsTheNumber committed:
Add the other armor pieces

FourIsTheNumber committed:
Try and add an item renderer

FourIsTheNumber committed:
Add tiers

FourIsTheNumber committed:
Begin file shenanigans

FourIsTheNumber committed:
Fix textures and frame mapping

FourIsTheNumber committed:
Backport full behavior system, implement night vision behavior

FourIsTheNumber committed:
Add augment items

FourIsTheNumber committed:
Add jetpack

FourIsTheNumber committed:
Add render system for augments

FourIsTheNumber committed:
Add fire immunity augment

FourIsTheNumber committed:
Add perfect hover and conditional augments

FourIsTheNumber committed:
Merge branch 'refs/heads/master' into armors

# Conflicts:
#	dependencies.gradle
#	src/main/resources/assets/gregtech/lang/en_US.lang

FourIsTheNumber committed:
Fix flight augment, add step assist augment

FourIsTheNumber committed:
Add goggles of revealing augment and questionable support for vis discount

FourIsTheNumber committed:
Decouple vis discount from behavior

FourIsTheNumber committed:
it's electric

FourIsTheNumber committed:
Add modeling... nightmare

FourIsTheNumber committed:
Dynamic model rendering based on augments, wip

MCTBL commented:
Why not just backport Construct's Armor from higher version?

FourIsTheNumber commented:
> Why not just backport Construct's Armor from higher version?

Because I want GT modular armor?

MCTBL commented:
> > Why not just backport Construct's Armor from higher version?
> 
> Because I want GT modular armor?

But back port CA can provide more armor options for early stages, and this Modular Armor maybe can like the upgrade from it.
Probably like player from steam stage move into electric stage. From CA move to MA?

FourIsTheNumber commented:
> But back port CA can provide more armor options for early stages, and this Modular Armor maybe can like the upgrade from it. Probably like player from steam stage move into electric stage. From CA move to MA?

Then someone else can backport that lol. That's not what I'm doing or what this pr is, I've already done most of the legwork to implement this system. I'm trying to replace IC2 electric armor with powered armor, whatever tinkers equivalent is way out of scope.

MCTBL commented:
> > But back port CA can provide more armor options for early stages, and this Modular Armor maybe can like the upgrade from it. Probably like player from steam stage move into electric stage. From CA move to MA?
> 
> Then someone else can backport that lol. That's not what I'm doing or what this pr is, I've already done most of the legwork to implement this system. I'm trying to replace IC2 electric armor with powered armor, whatever tinkers equivalent is way out of scope.

Maybe such a huge project need a voting? Like `areas-of-interest`?

GirixK commented:
> Maybe such a huge project need a voting? Like `areas-of-interest`?

If you find someone willing to work on backporting CA, they'll be free to contact one of the admins to open an AOI about it, until then it's a bit of a moot point

ReignOfFROZE commented:
Less important, but it would be _really_ cool if we could hide pieces of it like how you can with newer drac armor from 1.12 packs. I know personally I enjoy just seeing my skin/hiding just the helmet/boots/etc and if it's not a ton of trouble to implement, may be a nice QoL feature for others who feel the same. Obviously not something major that should block you shipping this feature, just an afterthought if you're bored :P

Dream-Master committed:
Merge branch 'master' into armors

spacebuilder2020 commented:
This reminds me of the Modular Powersuits mod! I think this is a very good idea, to have a GT modular armor

FourIsTheNumber committed:
Do some small tooltip and name updates

FourIsTheNumber committed:
Implement inertia canceling and setup omni mode

FourIsTheNumber committed:
Fix inertia canceling

FourIsTheNumber committed:
Improve lang system

FourIsTheNumber committed:
Add behaviors to aug tooltips

FourIsTheNumber committed:
Add energy usage

FourIsTheNumber committed:
Add hazmat

FourIsTheNumber committed:
Improve documentation

FourIsTheNumber committed:
Implement apiarist's armor, fix some deps

FourIsTheNumber committed:
Add swim speed augment

FourIsTheNumber committed:
Add knockback resist augment

FourIsTheNumber committed:
Add speed augment

FourIsTheNumber committed:
Add speed augment stacking and speed throttling

FourIsTheNumber committed:
Add jump boost augment

FourIsTheNumber committed:
Improve drain logic

FourIsTheNumber committed:
Remove testing stuff and old TODOs

FourIsTheNumber committed:
Add fall protection augment

FourIsTheNumber committed:
Add space suit augment (reliant on galaxyspace pr)

FourIsTheNumber committed:
Add missing behavior

FourIsTheNumber committed:
Add supported armor to the augment constructor

FourIsTheNumber committed:
Add localization and remove duplicate behavior from leggings

FourIsTheNumber committed:
Add infinite energy armor core

FourIsTheNumber committed:
Add force field augment

Dream-Master committed:
Merge branch 'master' into armors

FourIsTheNumber committed:
Add omni movement

FourIsTheNumber committed:
Add water breathing augment

FourIsTheNumber committed:
Add explanation and TODO for space suit nbt

FourIsTheNumber committed:
Fix waterbreathing

FourIsTheNumber committed:
New textures

FourIsTheNumber committed:
Merge branch 'refs/heads/master' into armors

# Conflicts:
#	dependencies.gradle
#	src/main/java/gregtech/api/util/GTUtility.java
#	src/main/resources/assets/gregtech/lang/en_US.lang

FourIsTheNumber committed:
Basic teleporter behavior

FourIsTheNumber committed:
Refactor augment items to use builder pattern, add foundation for category system

FourIsTheNumber committed:
Nightmare

FourIsTheNumber committed:
Use string ids for frames

YannickMG committed:
Quick & dirty working augment install/remove

FourIsTheNumber committed:
Change armor tag to use enum, automatically apply and remove augments instead of using submit button

FourIsTheNumber committed:
Add new slots, check for armor validity in slot filter

FourIsTheNumber committed:
Add colored borders to augment textures to show category

FourIsTheNumber committed:
Allow slotting frame and core, texture updates

FourIsTheNumber committed:
Frame textures and improvements

FourIsTheNumber committed:
Proper removal for stacked speed behavior

FourIsTheNumber committed:
Add core requirement implementation

FourIsTheNumber committed:
Add some chestplate augment overlays

FourIsTheNumber committed:
Add jetpack overlay and core requirements

FourIsTheNumber committed:
Enforce order of item removal

FourIsTheNumber committed:
Turn modification table from a basic machine to an MTE

FourIsTheNumber committed:
Remove *

FourIsTheNumber committed:
Require nano for space suit

FourIsTheNumber committed:
Swap cores around

FourIsTheNumber committed:
Add and balance frames

FourIsTheNumber committed:
Update textures with category borders

FourIsTheNumber committed:
Refactor away pointless classes

FourIsTheNumber committed:
Fix core tiering on textures and model

FourIsTheNumber committed:
Remove teleporter (extracted to new branch, still needs more ui work)

FourIsTheNumber committed:
Implement dynamic protection

FourIsTheNumber committed:
Work on energy tier

chrombread committed:
attempt to Merge remote-tracking branch 'origin/master' into armors

chrombread committed:
standardize renderer binding for armor

chrombread committed:
change behaviors to use new method

chrombread committed:
fix gui

chrombread committed:
change mod ids?

chrombread committed:
missed a couple method changes

chrombread committed:
i noobed lang

chrombread committed:
spotless for everynyan

chrombread committed:
move item slots into a column and use less .pos

FourIsTheNumber committed:
Merge remote-tracking branch 'origin/armors' into armors

FourIsTheNumber committed:
Fix isDown behavior

FourIsTheNumber committed:
Fix overcharging with core swaps

FourIsTheNumber committed:
Slightly improve texture cohesion

FourIsTheNumber committed:
Prevent removing dependency augments while dependent is slotted

FourIsTheNumber committed:
Add lang and category for keybinds

dagger8243 committed:
Modular Armor Charge PR (#5234)

FourIsTheNumber committed:
Add missing null check

FourIsTheNumber committed:
Merge remote-tracking branch 'origin/armors' into armors

FourIsTheNumber committed:
Show relevant keybinds on equip

FourIsTheNumber committed:
Textures for helmet augments

chrombread committed:
Merge remote-tracking branch 'origin/master' into armors

chrombread committed:
update dep

chrombread committed:
add temporary textures and interim frame bases for the armor

RecursivePineapple committed:
Merge branch 'master' into armors

RecursivePineapple committed:
Armor Behavior Refactor + Improvements (#5401)

FourIsTheNumber committed:
Merge branch 'refs/heads/master' into armors

FourIsTheNumber committed:
Merge branch 'refs/heads/master' into armors

# Conflicts:
#	src/main/java/gregtech/api/modularui2/GTGuiTextures.java

FourIsTheNumber committed:
Add missing null check

FourIsTheNumber committed:
Merge branch 'refs/heads/master' into armors

# Conflicts:
#	repositories.gradle
#	src/main/java/gregtech/api/enums/ItemList.java
#	src/main/java/gregtech/api/modularui2/GTGuiTextures.java
#	src/main/java/gregtech/api/modularui2/GTTextureIds.java
#	src/main/resources/assets/gregtech/lang/en_US.lang

FourIsTheNumber committed:
Merge

FourIsTheNumber commented:
This is now ready for testing. Will follow up with a coremod pr for recipes

Dream-Master committed:
Merge branch 'master' into armors

github-actions[bot] committed:
Optimising images contained in armors for #4065 (#5910)

Co-authored-by: FourIsTheNumber <33456283+FourIsTheNumber@users.noreply.github.com>
Co-authored-by: GitHub GTNH Actions <2970688+Dream-Master@users.noreply.github.com>

FourIsTheNumber committed:
Change frame items into augment items

Dream-Master committed:
Merge branch 'master' into armors

github-actions[bot] committed:
Optimising images contained in armors for #4065 (#5912)

Co-authored-by: FourIsTheNumber <33456283+FourIsTheNumber@users.noreply.github.com>
Co-authored-by: GitHub GTNH Actions <2970688+Dream-Master@users.noreply.github.com>

FourIsTheNumber committed:
Fix NPE

FourIsTheNumber committed:
Merge remote-tracking branch 'origin/armors' into armors

FourIsTheNumber committed:
Fix off-by-one core error

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

FourIsTheNumber committed:
Uptier jump and speed boost

FourIsTheNumber committed:
Merge remote-tracking branch 'origin/armors' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
fix

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
fix

github-actions[bot] committed:
Spotless apply for branch armors for #4065 (#6061)

Co-authored-by: GitHub GTNH Actions <>

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

leagris committed:
Merge branch 'master' into armors, resolved conflicts

Dream-Master committed:
Merge branch 'master' into armors

leagris committed:
Merge master into armors conflicts resolved

- GTGenericItem constructor now use 3 arguments
- Updated Item registration with new translation keys
- Moved inline English texts to language keys

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Eldrinn-Elantey commented:
I just tried flying in the dev environment and I have a question. Why does fall damage accumulate while flying and then reset when I land?

Dream-Master committed:
Merge branch 'master' into armors

ChangedWasHere commented:
How will this work with the thaumic boots mod?

would a "Meteor/Voidwalker/Comet" boot frame work?

Guvante commented on src/main/java/gregtech/api/hazards/HazardProtection.java:
```diff
@@ -45,21 +45,44 @@ public static boolean isWearingFullHazmatAgainst(@NotNull EntityLivingBase entit
+                return true;
+            }
+
+            if (!protectsAgainstHazard(entity, stack, hazard)) {
```
This shouldn't early out if a later protectsAgainstHazardFully item should allow it to succeed. You should be able to store a retVal that starts at true and this changes that to false finally returning that retVal if we finish the loop.

That way if you have a piece that doesn't before a protectsAgainstHazardFully it correctly checks the protectsAgainstHazardFully.

Guvante commented on src/main/java/gregtech/api/hazards/HazardProtection.java:
```diff
@@ -45,21 +45,44 @@ public static boolean isWearingFullHazmatAgainst(@NotNull EntityLivingBase entit
+        @NotNull Hazard hazard) {
+        return stack != null && (hasHazmatEnchant(stack) || (stack.getItem() instanceof IHazardProtector hazardProtector
+            && (hazardProtector.protectsAgainst(entity, stack, hazard)
+                || hazardProtector.protectsAgainstFully(entity, stack, hazard))));
```
This implies that a `hazardProtector` could return false for `protectsAgainst` but true for `protectsAgainstFully` is that the allowed behavior or is this just doing the safe thing and checking both?

Guvante commented on src/main/java/gregtech/api/items/armor/behaviors/CreativeFlightBehavior.java:
```diff
@@ -0,0 +1,58 @@
+
+        EntityPlayer player = context.getPlayer();
+
+        // Constantly re-check because alloyFlying is reset when you travel between dimensions
```
This is also necessary due to redundant behaviors removing the flag that is implicitly shard. But if it already needs to check each tick that is fine.

Guvante commented on src/main/java/gregtech/api/items/armor/behaviors/FireImmunityBehavior.java:
```diff
@@ -0,0 +1,55 @@
+    public boolean protectsAgainstFully(@NotNull ArmorContext context, Hazard hazard) {
+        // Protect against extreme temperatures.
+        // Protects fully - this behavior only needs to be on the leggings to work.
+        return switch (hazard) {
```
Just to double check, this can safely ignore the energy drain component, right?

Guvante commented on src/main/java/gregtech/api/items/armor/behaviors/InertiaCancelingBehavior.java:
```diff
@@ -0,0 +1,49 @@
+
+        EntityPlayer player = context.getPlayer();
+
+        if (player.moveForward == 0 && player.moveStrafing == 0 && player.capabilities.isFlying) {
```
Does this need to check the state of the behavior?

Guvante commented on src/main/java/gregtech/api/items/armor/behaviors/JetpackBehavior.java:
```diff
@@ -0,0 +1,108 @@
+                        player.motionY = Math.min(player.motionY + currentAccel, currentSpeedVertical);
+                    } else {
+                        if (descend) player.motionY = Math
+                            .min(player.motionY + currentAccel, jetpackStats.getVerticalHoverSlowSpeed());
```
The linked page has the hover slow speed negated

Guvante commented on src/main/java/gregtech/api/items/armor/ArmorEventHandlers.java:
```diff
@@ -0,0 +1,160 @@
+            }
+        }
+        if (player.stepHeight == MAGIC_STEP_HEIGHT) {
+            player.stepHeight = 0.6f;
```
Would it make sense to preserve and restore this value?

Guvante commented on src/main/java/gregtech/api/items/armor/ArmorEventHandlers.java:
```diff
@@ -0,0 +1,160 @@
+            if (jumpBoost > 0) {
+                player.motionY += jumpBoost;
+                player.fallDistance = player.fallDistance - (jumpBoost * 10);
+                context.drainEnergy(50);
```
So jumpBoost always takes energy but still works if none is available?

Guvante commented on src/main/java/gregtech/common/render/items/MechanicalArmorRenderer.java:
```diff
@@ -0,0 +1,88 @@
+        GTRenderUtil.renderItem(type, baseLayer);
+
+        if (state != null) {
+            // TODO: remove unnecessary second check thats just here to make my test world not crash
```
FYI Test code

Dream-Master committed:
Merge branch 'master' into armors

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/CreativeFlightBehavior.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public void onArmorTick(@NotNull ArmorContext context) {
+        if (!context.isRemote()) return;
```
CreativeFlightBehavior, SpeedBoostBehavior, and SwimSpeedBehavior all gate their onArmorTick behind `if (!context.isRemote()) return;`, so drainEnergy only runs client-side. The server never reduces charge for these behaviors, meaning the energy cost is never actually applied. Compare with JetpackBehavior which has no side check and drains correctly on both sides.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/JumpBoostBehavior.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public @NotNull IArmorBehavior merge(@NotNull IArmorBehavior other) {
+        if (!(other instanceof JumpBoostBehavior jumpBoost)) return null;
```
merge() returns null if other isn't a JumpBoostBehavior. ArmorState.addBehavior calls merge and puts the result into the map, so this will NPE or store null. Same issue in VisDiscountBehavior. Should probably return `this` on type mismatch instead.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/KnockbackResistBehavior.java:
```diff
@@ -0,0 +1,45 @@
+    protected KnockbackResistBehavior() {
+        attributes.put(
+            SharedMonsterAttributes.knockbackResistance.getAttributeUnlocalizedName(),
+            new AttributeModifier(UUID.randomUUID(), "KnockbackResistanceBehavior", 1, 0));
```
UUID.randomUUID() in a singleton constructor means the attribute modifier UUID changes every server restart. If the modifier is still on the player's attribute map from the previous session, the old one won't get cleaned up and the new one stacks on top. Use a fixed UUID.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/ArmorKeybinds.java:
```diff
@@ -0,0 +1,42 @@
+    public static final SyncedKeybind JETPACK_KEYBIND = SyncedKeybind
+        .createConfigurable("key.gt.toggle_jetpack", "Gregtech Armor", Keyboard.KEY_F);
+    public static final SyncedKeybind GOGGLES_OF_REVEALING_KEYBIND = SyncedKeybind
+        .createConfigurable("key.gt.toggle_goggles_of_revealing", "Gregtech Armor", Keyboard.KEY_T);
```
All three keybinds (goggles, inertia canceling, omnimovement) default to KEY_T. Players would need to rebind two of them before they can use them independently.

Luca-Guettinger commented on src/main/java/gregtech/common/items/armor/MechArmorBase.java:
```diff
@@ -0,0 +1,493 @@
+                            keyBind.getKeybinding()
+                                .getKeyDescription())
+                            + ": "
+                            + Keyboard.getKeyName(keyBind.getKeyCode()));
```
Keyboard.getKeyName() is an LWJGL client class. This runs on the server path (LivingEquipmentChangeEvent is server-only). It works in practice because lwjgl.jar is on the 1.7.10 server classpath and getKeyName is a pure array lookup, but it's fragile. Consider sending the key name from the client or resolving it from the SyncedKeybind config.

Luca-Guettinger commented on src/main/java/gregtech/common/items/armor/MechArmorBase.java:
```diff
@@ -0,0 +1,493 @@
+                    keyBind.registerPlayerListener(playerMP, this);
+
+                    if (!initMessage) {
+                        GTUtility.sendChatToPlayer(player, "Armor Systems Online... Active keybindings: ");
```
Hardcoded English string. Should use a lang key.

Luca-Guettinger commented on src/main/resources/assets/gregtech/lang/en_US.lang:
```diff
@@ -3323,6 +3323,119 @@ GT5U.item.pipe.amount=Pipe Amount
+GT5U.armor.part.tooltip.SpaceSuit=Adds Spacesuit capability. Must be applied to all four armor pieces
+GT5U.armor.part.tooltip.ForceField=Allows user to deploy a powerful forcefield which will block any damage for huge amounts of eu
+GT5U.armor.part.tooltip.OmniMovement=Allows user to move at increased speed in all directions
+GT5U.armor.part.tooltip.WaterBreathing=Restore's the user's air if they run out underwater
```
"Restore's" -> "Restores" (no apostrophe).

Luca-Guettinger reviewed (changes requested):
Client-side-only energy drain on creative flight, speed boost, and swim speed means the cost is never actually applied. JetpackBehavior shows the correct pattern (no side check). A few minor nits on top.

Dream-Master committed:
Merge branch 'master' into armors

boubou19 committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

donottellmetonottellyou commented:
I really like the idea of this, currently it is very easy to create god weapons using tinkers in HV and basically no HV armor can defend against it (1-shot). Are there any plans to balance/test this with Tinker's weapons?

Dream-Master committed:
Merge branch 'master' into armors

Ruling-0 commented:
Anyone working on this should check this PR (https://github.com/GTNewHorizons/GT5-Unofficial/pull/6590) targeting this branch.

Spaghetti-OberNub committed:
Fix: prevent jetpack movement while in GUI (#6590)

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Dream-Master committed:
Merge branch 'master' into armors

Eldrinn-Elantey committed:
Add HoloInventory augment for modular armor helmet

When installed and enabled via keybind (H), the armor acts as HoloInventory glasses by implementing IHoloGlasses on MechArmorBase and checking the active behavior state directly in NBT to avoid full ArmorState deserialization on every render frame.

Eldrinn-Elantey committed:
fix armor behavior save, final INSTANCE fields, i18n

ForceFieldBehavior and OmniMovementBehavior were calling ArmorState.toggle() directly, skipping save() — active state would reset on relog. Fixed by using context.toggleBehavior(). Added missing final modifier to INSTANCE fields in FireImmunityBehavior, InertiaCancelingBehavior, JetpackPerfectHoverBehavior, MilkInfusionBehavior, OmniMovementBehavior, ForceFieldBehavior. Replaced StatCollector.translateToLocal with GTUtility.translate in VisDiscountBehavior. Replaced hardcoded English strings in ArmorState tooltip with lang keys, added missing HoloInventory augment tooltip.

Eldrinn-Elantey committed:
fix review comments and bugs for modular armor system

* fix creative flight, speed boost, swim speed draining energy client-side only; drain now runs on both sides
* fix creative flight not setting allowFlying on server, causing flying to be rejected on dedicated servers
* fix fall damage accumulating while creative flight is active
* fix jetpack triggering when pressing spacebar in inventory GUI
* fix ForceField and OmniMovement toggle state not persisting across relogs
* fix InertiaCanceling toggle state not persisting across relogs
* fix InertiaCanceling tick not checking whether behavior is toggled on
* fix JetpackBehavior hover+descend speed not negated, causing upward push instead of downward
* fix JumpBoost and VisDiscount merge() returning null on type mismatch, causing NPE
* fix KnockbackResist applying stacking attribute modifiers across restarts due to random UUID in singleton
* fix HazardProtection early-returning before checking protectsAgainstFully on subsequent slots
* fix jumpBoost applying even when boots have no energy
* fix hardcoded English strings in ArmorState tooltips and onArmorEquip message; add lang keys
* fix Keyboard.getKeyName() call crashing on dedicated server if keybinding is null
* fix "Restore's" typo in WaterBreathing tooltip
* change InertiaCanceling default keybind from T to J, OmniMovement from T to V to resolve conflicts
* preserve and restore original stepHeight when StepAssist is removed

Eldrinn-Elantey committed:
Update dependencies.gradle

Eldrinn-Elantey committed:
Merge branch 'master' into armors

github-actions[bot] committed:
Optimising images contained in armors for #4065 (#6759)

Co-authored-by: FourIsTheNumber <33456283+FourIsTheNumber@users.noreply.github.com>
Co-authored-by: GitHub GTNH Actions <46845681+Eldrinn-Elantey@users.noreply.github.com>

Guvante commented on src/main/java/gregtech/api/items/armor/behaviors/CreativeFlightBehavior.java:
```diff
@@ -29,23 +29,19 @@ public IIcon getModularArmorTexture() {
-            player.capabilities.allowFlying = true;
-            player.sendPlayerAbilities();
-        }
+        boolean hasEnergy = context.drainEnergy(75);
```
Does this drain the energy even when not flying?

Guvante commented on src/main/java/gregtech/api/items/armor/ArmorEventHandlers.java:
```diff
@@ -140,6 +145,15 @@ public void onPlayerJump(LivingEvent.LivingJumpEvent event) {
+            ItemStack chest = player.getCurrentArmor(SLOT_CHEST);
+            if (chest != null && chest.getItem() instanceof MechArmorBase) {
+                ArmorContext chestContext = MechArmorBase.load(player, chest);
+                if (chestContext.hasBehavior(BehaviorName.CreativeFlight)) {
```
Does a drained chest piece have this behavior? Or is cancelling fall damage without power an intended benefit here.

Guvante commented on src/main/java/gregtech/api/items/armor/ArmorState.java:
```diff
@@ -0,0 +1,354 @@
+            }
+        }
+
+        if (showAllInfo) {
```
Redundant if could combine with the previous one

Dream-Master committed:
Merge branch 'master' into armors

Luca-Guettinger commented on src/main/java/gregtech/common/tileentities/machines/basic/MTEModificationTable.java:
```diff
@@ -0,0 +1,456 @@
+        setArmorState(state);
+    }
+
+    LimitingItemStackHandler armorSlotHandler = new LimitingItemStackHandler(3, 1);
```
armorSlotHandler and augmentsSlotHandler are not connected to mInventory. BlockMachines.breakBlock iterates mInventory to drop items, so everything in these custom handlers gets voided when the block is broken. setItemNBT is not overridden either, so the dropped machine block won't carry the data. This permanently loses the player's armor piece and all installed components.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/SpeedBoostBehavior.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public @NotNull IArmorBehavior merge(@NotNull IArmorBehavior other) {
```
Unchecked cast in merge(), same issue as JumpBoostBehavior. SwimSpeedBehavior.merge has it too.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/SpeedBoostBehavior.java:
```diff
@@ -0,0 +1,122 @@
+
+        state.speedBoost = MathUtils.clamp(state.speedBoost, 0, 1);
+
+        GTUtility.sendChatToPlayer(context.getPlayer(), "New speed: " + Math.round(state.speedBoost * 100F) + "%");
```
Hardcoded English string. Should use a lang key like the other chat messages.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/WaterBreathingBehavior.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public void onArmorTick(@NotNull ArmorContext context) {
```
Missing side check. NightVisionBehavior, FireImmunityBehavior, MilkInfusionBehavior all have if (context.isRemote()) return; for server-side operations. Air replenishment and energy drain should also be server-only.

Luca-Guettinger commented on src/main/java/gregtech/common/GTProxy.java:
```diff
@@ -1063,11 +1065,17 @@ public void onInitialization(FMLInitializationEvent event) {
         if (!this.enableUndergroundGravelGen) PREVENTED_ORES.add(OreGenEvent.GenerateMinable.EventType.GRAVEL);
         if (!this.enableUndergroundDirtGen) PREVENTED_ORES.add(OreGenEvent.GenerateMinable.EventType.DIRT);
 
+        ArmorEventHandlers ArmorEvents = new ArmorEventHandlers();
```
ArmorEvents -> armorEvents

Luca-Guettinger reviewed (changes requested):
1 item loss bug in the modification table, 4 minor things.

Luca-Guettinger committed:
Merge branch 'master' into armors

Eldrinn-Elantey committed:
fix review comments

- drain energy in CreativeFlightBehavior only when player is actually flying
- check canDrainEnergy before canceling fall damage from CreativeFlight, add canDrainEnergy helper to ArmorState
- merge duplicate if (showAllInfo) blocks in ArmorState.addArmorInformation
- fix MTEModificationTable dropping armor on break: sync armorSlotHandler slot 0 to mInventory[0], override setItemNBT to persist armor into dropped block item
- guard SpeedBoostBehavior.merge and SwimSpeedBehavior.merge with instanceof check instead of unchecked cast
- replace hardcoded "New speed: ..." string in SpeedBoostBehavior with GT5U.armor.message.speed_set lang key
- add server-side guard in WaterBreathingBehavior.onArmorTick
- rename ArmorEvents variable to armorEvents in GTProxy

Eldrinn-Elantey committed:
Merge branch 'armors' of https://github.com/GTNewHorizons/GT5-Unofficial into armors

Eldrinn-Elantey committed:
typo

Eldrinn-Elantey committed:
Merge branch 'master' into armors

Eldrinn-Elantey committed:
armor stays in modification table slot when GUI is closed; drop it near the table on close instead. uses a one-tick delay via onPostTick to let ModularUI finish syncing before spawning the item, then sends inventoryContainer to the client to clear ghost items

Luca-Guettinger committed:
Merge branch 'master' into armors

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/ArmorEventHandlers.java:
```diff
@@ -0,0 +1,175 @@
+                toTag.removeTag("charge");
+                fromTag.removeTag("charge");
+
+                if (GTUtility.areStacksEqual(from, toCopy)) return;
```
areStacksEqual(from, toCopy) compares the original from (still has the charge tag) against toCopy (charge removed). This never matches when charge differs, so every tick that save() writes new charge to NBT triggers the full equip/unequip path: chat spam, keybind re-registration, behavior callbacks. Should be areStacksEqual(fromCopy, toCopy).

Luca-Guettinger commented on src/main/java/gregtech/common/items/armor/MechArmorBase.java:
```diff
@@ -0,0 +1,510 @@
+    @SideOnly(Side.CLIENT)
+    public ModelBiped getArmorModel(EntityLivingBase entityLiving, ItemStack itemStack, int armorSlot) {
+        ModelMechArmor model;
+        if (armorSlot == SLOT_LEGS) model = new ModelMechArmor(0.25F);
```
getArmorModel allocates a new ModelMechArmor on every call. This runs per render frame per armor piece. Cache the two model instances and reconfigure showModel flags on them instead.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/behaviors/NightVisionBehavior.java:
```diff
@@ -0,0 +1,68 @@
+                .removePotionEffect(Potion.blindness.id);
+            context.getPlayer()
+                .addPotionEffect(new PotionEffect(Potion.nightVision.id, 999999, 0, true));
+        } else {
```
The else branch strips night vision unconditionally, including NV from potions or other mods. If the behavior is toggled off, external NV gets removed every tick. Should only remove if the behavior was previously providing it.

Luca-Guettinger commented on src/main/java/gregtech/common/items/armor/MechArmorBase.java:
```diff
@@ -0,0 +1,510 @@
+            behavior.onArmorTick(context);
+        }
+
+        context.save();
```
context.save() runs unconditionally every tick for every armor piece, rewriting the entire NBT compound each call. A dirty flag on ArmorContext that's set by drainEnergy/toggleBehavior/etc. would let you skip the write when nothing changed.

Luca-Guettinger commented on src/main/resources/assets/gregtech/lang/en_US.lang:
```diff
@@ -3467,6 +3467,131 @@ GT5U.item.pipe.amount=Pipe Amount
+GT5U.armor.part.tooltip.JumpBoost=Allows the user to jump higher
+GT5U.armor.part.tooltip.FallProtection=Converts the user's fall damage into EU usage
+GT5U.armor.part.tooltip.SpaceSuit=Adds Spacesuit capability. Must be applied to all four armor pieces
+GT5U.armor.part.tooltip.ForceField=Allows user to deploy a powerful forcefield which will block any damage for huge amounts of eu
```
"eu" -> "EU"

Luca-Guettinger reviewed (changes requested):
some more things i found

Eldrinn-Elantey committed:
fix armor tick spam, per-frame model alloc, and NV potion conflict from review

Eldrinn-Elantey committed:
move keybind chat messages to client side, remove test comment in renderer

Dream-Master committed:
Merge branch 'master' into armors

Luca-Guettinger commented on src/main/java/gregtech/common/items/armor/MechArmorBase.java:
```diff
@@ -0,0 +1,527 @@
+    }
+
+    @Override
+    public void damageArmor(EntityLivingBase entity, ItemStack stack, DamageSource source, int damage, int slot) {
```
damageArmor drains energy but never calls context.save(). The charge deduction is lost on the next load. Compare with getProperties at line 352 which correctly saves after drainEnergy. Armor damage EU cost is effectively zero right now.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/ArmorState.java:
```diff
@@ -0,0 +1,356 @@
+                if (categoryTag == null) continue;
+
+                // noinspection unchecked
+                for (var e : ((Map<String, NBTTagString>) categoryTag.tagMap).entrySet()) {
```
Direct field access to tagMap with unchecked cast to Map<String, NBTTagString>. A non-string tag in this compound (NBT editing, format change) would ClassCastException at iteration time. getKeySet() + getString() is safer.

Luca-Guettinger commented on src/main/java/gregtech/api/items/armor/ArmorState.java:
```diff
@@ -0,0 +1,356 @@
+
+        NBTTagCompound augmentTag = tag.getCompoundTag("augments");
+
+        if (augmentTag != null) {
```
getCompoundTag never returns null (returns empty compound if missing). The null check here and the categoryTag == null continue on line 243 are dead code.

Luca-Guettinger commented on src/main/java/gregtech/common/tileentities/machines/basic/MTEModificationTable.java:
```diff
@@ -0,0 +1,505 @@
+            if (!NetworkUtils.isClient(player)) {
+                ItemStack armor = armorSlotHandler.getStackInSlot(0);
+                if (armor != null) {
+                    armorSlotHandler.setStackInSlot(0, null);
```
Armor is nulled in the handler and stored in pendingArmorDrop (transient). If the server saves between close and the next tick, saveNBTData writes null and pendingArmorDrop is gone. Give it to the player's inventory directly instead of deferring to onPostTick.

Luca-Guettinger reviewed (changes requested):
1 energy persistence bug in damageArmor, 3 smaller things.

Eldrinn-Elantey committed:
fix armor charge not saved on damage, unsafe NBT cast, and armor loss on server save

Luca-Guettinger committed:
small fix

Luca-Guettinger committed:
Merge branch 'master' into armors

Luca-Guettinger committed:
Merge branch 'master' into armors
